File: ZalaBackend/app/schemas/lead.py
from typing import Optional, List

from pydantic import BaseModel, Field

from app.schemas.address import AddressPublic
from app.schemas.contact import ContactPublic
from app.schemas.summaries import UserSummary
from app.schemas.property import PropertyPublic
from app.schemas.lead_image import LeadImagePublic

class LeadBase(BaseModel):
    """
    Base Schema for a Lead.
    """
    person_type: Optional[str] = None
    business: Optional[str] = None
    website: Optional[str] = None
    license_num: Optional[str] = None
    notes: Optional[str] = None
    image_url: Optional[str] = None


class LeadCreate(LeadBase):
    """
    Schema for Create a Lead.
    """
    # contact: ContactBase
    # address: Optional[AddressBase] = None
    # Linking a user to a lead should be done via the link endpoint
    # POST /leads/{lead_id}/users/{user_id} and not via the create body.


class LeadUpdate(BaseModel):
    """
    Schema for Updating a Lead
    """
    # contact: Optional[ContactBase] = None
    # address: Optional[AddressUpdate] = None
    person_type: Optional[str] = None
    business: Optional[str] = None
    website: Optional[str] = None
    license_num: Optional[str] = None
    notes: Optional[str] = None
    image_url: Optional[str] = None


class LeadPublic(LeadBase):
    """
    Schema for Get a Lead
    """
    lead_id: int

    # expose ids for related resources to keep response small and avoid circular imports
    created_by: Optional[int] = None
    contact_id: Optional[int] = None
    address_id: Optional[int] = None

    # nested created_by user summary and full contact/property details
    created_by_user: Optional[UserSummary] = None
    contact: Optional[ContactPublic] = None
    address: Optional[AddressPublic] = None
    properties: List[PropertyPublic] = []
    campaigns: List["CampaignLeadPublic"] = []
    images: List[LeadImagePublic] = Field(default_factory=list)

    class Config:
        from_attributes = True

File: ZalaFrontend/src/interfaces/Lead.ts
import { AAddressToIAddress, type AAddress, type IAddress } from "./Address";
import { type AContact, type IContact, AContactToIContact } from "./contact";
import { AImageToIImage, type AImage, type IImage } from "./Image";
import { AUserToIUser, type AUser, type IUser } from "./user";

export type ALead = {
  lead_id: number;
  created_by: number;
  contact_id: number;
  address_id: number;

  person_type: "person" | "business"; // could extend if needed

  business: string;
  website: string;
  license_num: string;
  notes: string;

  contact: AContact;
  address: AAddress;

  image_url?: string | null;
  images?: AImage[];

  properties: never[]; // Not sure these are needed in response from api
  campaigns: never[]; // Not sure these are needed in response from api

  created_by_user: AUser;
};

export type ILead = {
  personType: "person" | "business";
  leadId: number;

  buisness: string;
  website: string;
  licenseNum: string;
  notes: string;

  contact?: IContact;
  address?: IAddress;

  imageUrl?: string;
  images?: IImage[];

  createdBy?: number;
  createdByUser?: IUser;
};

export const ALeadToILead = (body: ALead): ILead => ({
  leadId: body.lead_id,
  personType: body.person_type ?? "person",
  buisness: body.business ?? "",
  website: body.website ?? "",
  licenseNum: body.license_num ?? "",
  notes: body.notes ?? "",
  imageUrl: body.image_url ?? undefined,
  images: body.images ? body.images.map(AImageToIImage) : undefined,
  createdBy: body.created_by ?? undefined,
  createdByUser: body.created_by_user
    ? AUserToIUser(body.created_by_user)
    : undefined,
  contact: body.contact ? AContactToIContact(body.contact) : undefined,
  address: body.address ? AAddressToIAddress(body.address) : undefined,
});
